### PR TITLE
Static initialization

### PR DESCRIPTION
**Pull request by masterleinad**
This PR attempts to avoid problems with the initialization order of global variables as much as possible by deferring initialization upon first use or initializing the variables as compile time constants.
This PR does not change any of the static public variables (although that would probably be even more beneficial).

Apart from that:
- `clang-tidy` is run on `expand_instantiations` and
- `Points` can be created as compile-time constants.

### CODE REVIEW

File: include/deal.II/base/event.h [hunk 1]
```diff
@@ -134,7 +134,8 @@ namespace Algorithms
      * The names of registered events
      */
     // TODO: This static field must be guarded by a mutex to be thread-safe!
-    static std::vector<std::string> names;
+    static std::vector<std::string> &
+    get_names();
```
Comment (bangerth) [inline include/deal.II/base/event.h]: The TODO comment now no longer applies properly. Can you update it?

File: include/deal.II/base/job_identifier.h [hunk 2]
```diff
@@ -66,21 +66,19 @@ class JobIdentifier
...
 };
 
-
-/*----------------------------- Inline functions ----------------------------*/
-
-
-/**
- * Global object to identify the presently running program.
- */
-extern JobIdentifier dealjobid;
```
Comment (bangerth) [inline include/deal.II/base/job_identifier.h]: this is strictly speaking an incompatible change. I think it's so obscure that there isn't even a point in mentioning that :-)

File: include/deal.II/base/multithread_info.h [hunk 3]
```diff
@@ -117,23 +117,10 @@ class MultithreadInfo
...
-   */
-  static unsigned int n_max_threads;
-
-  /**
-   * Variable representing the number of cores in the system. This is computed
-   * by get_n_cpus() and is returned by n_cores().
-   */
-  static const unsigned int n_cpus;
+  static unsigned int &
+  get_n_max_threads();
```
Comment (bangerth) [inline include/deal.II/base/multithread_info.h]: Why is this necessary? Static initialization of `unsigned int` variables is pretty uncritical (even if it involves a function call) because it happens before `main()` even starts.

File: include/deal.II/base/point.h [hunk 4]
```diff
@@ -338,7 +338,9 @@ inline Point<dim, Number>::Point(const Number x, const Number y)
 
 
 template <int dim, typename Number>
-inline Point<dim, Number>::Point(const Number x, const Number y, const Number z)
+inline constexpr Point<dim, Number>::Point(const Number x,
+                                           const Number y,
+                                           const Number z)
```
Comment (drwells) [inline include/deal.II/base/point.h]: Don't we have problems with `Assert` and `constexpr` with intel, i.e., won't this just cause the same problems as we had in #7134?

**Comment by masterleinad** [comment]
> I suspect not all necessary constexpr functions are available in C++11.

That's sad :cry:. I will try to find another solution then.

**Comment by masterleinad** [comment]
> Here are three comments, but I haven't really tried to read all the way through because there are just too many independent things in one patch. Can you break it apart? For example one pull request for the whitespace changes on `expand_instantiations`, one for the `constexpr` changes to `Point`, and one for the static variables?

Sure. I will create some separate PRs.

**Comment by masterleinad** [comment]
I just dropped the `constexpr` support for `Point`. It is just too annoying to get that working at all with just C++11.

File: include/deal.II/base/job_identifier.h (hunk 2 above)
Comment (masterleinad) [reply include/deal.II/base/job_identifier.h]: Yes, you are right, of course.

File: include/deal.II/base/multithread_info.h (hunk 3 above)
Comment (masterleinad) [reply include/deal.II/base/multithread_info.h]: Well, all static initialization of global variables happens before `main`, but the initialization order is undefined. The whole PR is about this at the end whether it probable to be a problem or not.

I guess you are right that it is really hard to run into an error here though...

**Comment by masterleinad** [comment]
I reduced the size of changes here by only lazily initialize non-class static variables (possibly accessible from other compilation unit's static variables).
The only private static member variable whose destructor has a non-negligible side-effect to the environment is `GrowingVectorMemory::Pool` which is handled in #7448.

`GridGenerator`'s static global variables were actually guarded by an unnamed namespace, but I just decided to move them to the only place where they are actually used.

**Comment by masterleinad** [comment]
This is the lists of warnings left:
```
../examples/step-14/step-14.cc:1097:63: warning: declaration requires a global destructor [-Wglobal-constructors]
../examples/step-14/step-14.cc:1099:62: warning: declaration requires a global destructor [-Wglobal-constructors]
../examples/step-14/step-14.cc:1638:57: warning: declaration requires a global destructor [-Wglobal-constructors]
../examples/step-18/step-18.cc:701:48: warning: declaration requires a global constructor [-Wglobal-constructors]
../examples/step-19/step-19.cc:55:28: warning: declaration requires a global destructor [-Wglobal-constructors]
../examples/step-19/step-19.cc:56:28: warning: declaration requires a global destructor [-Wglobal-constructors]
../examples/step-19/step-19.cc:57:28: warning: declaration requires a global destructor [-Wglobal-constructors]
../examples/step-19/step-19.cc:58:28: warning: declaration requires a global destructor [-Wglobal-constructors]
../examples/step-21/step-21.cc:379:22: warning: declaration requires a global destructor [-Wglobal-constructors]
../examples/step-39/step-39.cc:76:41: warning: declaration requires a global destructor [-Wglobal-constructors]
../examples/step-43/step-43.cc:260:22: warning: declaration requires a global destructor [-Wglobal-constructors]
../examples/step-51/step-51.cc:109:22: warning: declaration requires a global constructor [-Wglobal-constructors]
../examples/step-51/step-51.cc:115:22: warning: declaration requires a global constructor [-Wglobal-constructors]
../examples/step-51/step-51.cc:120:22: warning: declaration requires a global constructor [-Wglobal-constructors]
../examples/step-7/step-7.cc:125:50: warning: declaration requires a global constructor [-Wglobal-constructors]
../examples/step-7/step-7.cc:131:50: warning: declaration requires a global constructor [-Wglobal-constructors]
../examples/step-9/step-9.cc:165:36: warning: declaration requires a global constructor [-Wglobal-constructors]
../examples/step-9/step-9.cc:168:36: warning: declaration requires a global constructor [-Wglobal-constructors]
../examples/step-9/step-9.cc:171:36: warning: declaration requires a global constructor [-Wglobal-constructors]
../source/base/event.cc:25:35: warning: declaration requires a global destructor [-Wglobal-constructors]
../source/base/event.cc:65:17: warning: declaration requires a global destructor [-Wglobal-constructors]
../source/base/event.cc:66:17: warning: declaration requires a global destructor [-Wglobal-constructors]
../source/base/event.cc:67:17: warning: declaration requires a global destructor [-Wglobal-constructors]
../source/base/event.cc:68:17: warning: declaration requires a global destructor [-Wglobal-constructors]
../source/base/event.cc:69:17: warning: declaration requires a global destructor [-Wglobal-constructors]
../source/base/logstream.cc:37:11: warning: declaration requires a global destructor [-Wglobal-constructors]
../source/base/multithread_info.cc:232:37: warning: declaration requires a global constructor [-Wglobal-constructors]
../source/base/multithread_info.cc:246:5: warning: declaration requires a global constructor [-Wglobal-constructors]
../source/base/parameter_acceptor.cc:28:65: warning: declaration requires a global destructor [-Wglobal-constructors]
../source/base/parameter_acceptor.cc:30:37: warning: declaration requires a global destructor [-Wglobal-constructors]
../source/base/path_search.cc:28:61: warning: declaration requires a global destructor [-Wglobal-constructors]
../source/base/path_search.cc:29:61: warning: declaration requires a global destructor [-Wglobal-constructors]
../source/base/path_search.cc:30:61: warning: declaration requires a global destructor [-Wglobal-constructors]
../source/base/polynomial.cc:989:19: warning: declaration requires a global destructor [-Wglobal-constructors]
../source/fe/mapping_q1.cc:66:35: warning: declaration requires a global destructor [-Wglobal-constructors]
../source/hp/mapping_collection.cc:100:37: warning: declaration requires a global destructor [-Wglobal-constructors]
../source/lac/affine_constraints.cc:149:3: warning: declaration requires a global destructor [-Wglobal-constructors]
../source/lac/affine_constraints.cc:150:3: warning: declaration requires a global destructor [-Wglobal-constructors]
../source/numerics/time_dependent.cc:1191:50: warning: declaration requires a global destructor [-Wglobal-constructors]
../source/physics/elasticity/standard_tensors.cc:26:46: warning: declaration requires a global constructor [-Wglobal-constructors]
../source/physics/elasticity/standard_tensors.cc:32:46: warning: declaration requires a global constructor [-Wglobal-constructors]
../source/physics/elasticity/standard_tensors.cc:37:74: warning: declaration requires a global constructor [-Wglobal-constructors]
../source/physics/elasticity/standard_tensors.cc:44:46: warning: declaration requires a global constructor [-Wglobal-constructors]
```

File: source/grid/grid_generator.cc [hunk 5]
```diff
@@ -3919,6 +3899,16 @@ namespace GridGenerator
...
 
+    // Corner points of the cube [-1,1]^3
+    static const std::array<Point<3>, 8> hexahedron = {{{-1, -1, -1}, //
+                                                        {+1, -1, -1}, //
+                                                        {-1, +1, -1}, //
+                                                        {+1, +1, -1}, //
+                                                        {-1, -1, +1}, //
+                                                        {+1, -1, +1}, //
+                                                        {-1, +1, +1}, //
+                                                        {+1, +1, +1}}};
```
Comment (drwells) [inline source/grid/grid_generator.cc]: Would it be possible to mark this as `constexpr`, or would that require C++14?

Comment (masterleinad) [reply source/grid/grid_generator.cc]: I couldn't find a (suitably simple) way to make the `Point` constructor used here `constexpr` using just C++11. That requires to make `Tensor` `constexpr` which is not really easy due to all our customization for specific `Number` template parameters.

Comment (drwells) [reply source/grid/grid_generator.cc]: That's right; I forgot about that.

File: cmake/scripts/expand_instantiations.cc [hunk 6]
```diff
@@ -52,11 +52,16 @@
...
 
-// a map from the keys in the expansion lists to the list itself. For
+// Returns a map from the keys in the expansion lists to the list itself. For
 // instance, the example above will lead to the entry
-//      expansion_lists[REAL_SCALARS] = (double, float)
+//      get_expansion_lists()[REAL_SCALARS] = (double, float)
 // in this map, among others
-std::map<std::string, std::list<std::string>> expansion_lists;
+std::map<std::string, std::list<std::string>> &
+get_expansion_lists()
```
Comment (bangerth) [inline cmake/scripts/expand_instantiations.cc]: I'm not opposed to this change, but I question the wisdom of the warning. We are here in a user program that depends on libstdc++, and we are constructing an object that only depends on libstdc++ functionality. Since libstdc++ is in a shared lib, all of its global variable constructors run before we even get to `main()`, and consequently by the time we run the constructor of the global variable that was here in the existing code, everything is in fact already constructed and ready to use. There is no ordering question, though it is conceivable that there is an ordering problem with other global variables in this file.

I'm ok with the change in question, primarily because I'm in favor of keeping code uniform. I'm just questioning whether that flag is as useful...

File: include/deal.II/fe/fe_tools.templates.h [hunk 7]
```diff
@@ -1303,11 +1303,19 @@ namespace FETools
       // running, there are no thread-safety issues here. since this is
       // compiled for all dimensions at once, need to create objects for
       // each dimension and then separate between them further down
-      static std::array<
+      std::array<
```
Comment (bangerth) [inline include/deal.II/fe/fe_tools.templates.h]: you're in a header file -- I think you need to mark this function as `inline`.

File: cmake/scripts/expand_instantiations.cc (hunk 6 above)
Comment (masterleinad) [reply cmake/scripts/expand_instantiations.cc]: Are you fine with the PR if I drop this commit?

Comment (bangerth) [reply cmake/scripts/expand_instantiations.cc]: Let's not worry about it -- it's fine as is!
